File: dao-style-text/src/main/java/example/dao_style_text/entity/EmployeeListener.java
package example.dao_style_text.entity;

import java.sql.Timestamp;
import org.seasar.doma.jdbc.entity.EntityListener;
import org.seasar.doma.jdbc.entity.PostDeleteContext;
import org.seasar.doma.jdbc.entity.PostInsertContext;
import org.seasar.doma.jdbc.entity.PostUpdateContext;
import org.seasar.doma.jdbc.entity.PreDeleteContext;
import org.seasar.doma.jdbc.entity.PreInsertContext;
import org.seasar.doma.jdbc.entity.PreUpdateContext;

public class EmployeeListener<E extends Employee> implements EntityListener<E> {

  @Override
  public void preDelete(E employee, PreDeleteContext<E> context) {}

  @Override
  public void preInsert(E employee, PreInsertContext<E> context) {
    var timestamp = new Timestamp(System.currentTimeMillis());
    employee.setInsertTimestamp(timestamp);
  }

  @Override
  public void preUpdate(E employee, PreUpdateContext<E> context) {
    if (context.isEntityChanged()) {
      var timestamp = new Timestamp(System.currentTimeMillis());
      employee.setUpdateTimestamp(timestamp);
    }
  }

  @Override
  public void postInsert(E entity, PostInsertContext<E> context) {}

  @Override
  public void postUpdate(E entity, PostUpdateContext<E> context) {}

  @Override
  public void postDelete(E entity, PostDeleteContext<E> context) {}
}

File: dsl-style-kotlin/src/main/kotlin/example/dsl_style_kotlin/entity/EmployeeListener.kt
package example.dsl_style_kotlin.entity

import org.seasar.doma.jdbc.entity.EntityListener
import org.seasar.doma.jdbc.entity.PostDeleteContext
import org.seasar.doma.jdbc.entity.PostInsertContext
import org.seasar.doma.jdbc.entity.PostUpdateContext
import org.seasar.doma.jdbc.entity.PreDeleteContext
import org.seasar.doma.jdbc.entity.PreInsertContext
import org.seasar.doma.jdbc.entity.PreUpdateContext
import java.sql.Timestamp

class EmployeeListener : EntityListener<Employee> {
    override fun preDelete(employee: Employee, context: PreDeleteContext<Employee>) {}
    override fun preInsert(employee: Employee, context: PreInsertContext<Employee>) {
        val timestamp = Timestamp(System.currentTimeMillis())
        employee.insertTimestamp = timestamp
    }

    override fun preUpdate(employee: Employee, context: PreUpdateContext<Employee>) {
        if (context.isEntityChanged) {
            val timestamp = Timestamp(System.currentTimeMillis())
            employee.updateTimestamp = timestamp
        }
    }

    override fun postInsert(entity: Employee, context: PostInsertContext<Employee>) {}
    override fun postUpdate(entity: Employee, context: PostUpdateContext<Employee>) {}
    override fun postDelete(entity: Employee, context: PostDeleteContext<Employee>) {}
}
